lucille-core/src/main/java/com/kmwllc/lucille/stage/EmbeddedPythonStage.java

```diff
@@ -0,0 +1,671 @@
+  private Context context;
+  private Engine engine;
+
+  public EmbeddedPythonStage(Config config) {
+    super(config);
+    this.scriptPath = config.hasPath("script_path") ? config.getString("script_path") : null;
+    this.inlineScript = config.hasPath("script") ? config.getString("script") : null;
+//    this.pythonRoot = config.hasPath("python_root") ? config.getString("python_root") : null;
+  }
+
+  @Override
+  public void start() throws StageException {
+    if ((scriptPath == null && inlineScript == null) || (scriptPath != null && inlineScript != null)) {
+      throw new StageException("Can only specify either script path or script.");
+    }
+
+    if (inlineScript != null) {
+      try {
+        this.source = Source.newBuilder("python", inlineScript, "<inline>").build();
+      } catch (Exception e) {
+        throw new StageException("Failed to build inline Python source.", e);
+      }
+    } else {
+      try (Reader reader = FileContentFetcher.getOneTimeReader(scriptPath, config)) {
+        this.source = Source.newBuilder("python", reader, scriptPath).build();
+      } catch (Exception e) {
+        throw new StageException("Failed to read Python from '" + scriptPath + "'.", e);
+      }
+    }
+
+    try {
+      this.engine = Engine.newBuilder()
+          .option("engine.WarnInterpreterOnly", "false")
+          .option("log.level", "OFF")
+          .build();
+
+      this.context = Context.newBuilder("python")
+          .engine(engine)
+          .allowAllAccess(true)
+          .build();
+
+      // Code for Python dependencies, but requires JDK 21
+
+//      if (pythonRoot != null) {
+//        Path pyRoot = Paths.get(pythonRoot);
+//        this.context = GraalPyResources.contextBuilder(pyRoot)
+//            .engine(engine)
+//            .out(System.out)
+//            .err(System.err)
+//            .build();
+//      } else {
+//        this.context = Context.newBuilder("python")
+//            .engine(engine)
+//            .allowAllAccess(true)
+//            .out(System.out)
+//            .err(System.err)
+//            .build();
+//      }
+    } catch (Exception e) {
+      throw new StageException("Failed to initialize Python context.", e);
+    }
+  }
+
+  @Override
+  public void stop() throws StageException {
+    if (context != null) {
+      context.close();
+    }
+
+    if (engine != null) {
+      engine.close();
+    }
+  }
+
+  @Override
+  public Iterator<Document> processDocument(Document doc) throws StageException {
+    try {
+      context.getBindings("python").putMember("doc", PyDocProxy.root(doc));
+      context.getBindings("python").putMember("rawDoc", doc);
+      context.eval(source);
+      return null;
+    } catch (Exception e) {
+      throw new StageException("Python failed for doc '" + doc.getId() + "' using '" + scriptPath + "'.", e);
+    }
+  }
+
+  // Proxy implementation exposed to Python
+
+  private static final class PyDocProxy implements ProxyObject, ProxyHashMap {
```

Comment:
it looks like we are supporting two paradigms where the python doc can be treated as a dictionary (accessing fields via `doc['name']` or as an object where the fields are properties that you access via `doc.name`

I think it's OK to move forward supporting both but we should consider whether this might be confusing for a python developer who isn't expecting these paradigms to be combined?

in any case, we should include some documentation here about what we decided and why